src/mono/mono/mini/simd-intrinsics.c

```diff
@@ -2732,12 +2731,22 @@ emit_vector_2_3_4 (MonoCompile *cfg, MonoMethod *cmethod, MonoMethodSignature *f
 		g_assert (fsig->hasthis && fsig->param_count == 2 && fsig->params [0]->type == MONO_TYPE_I4 && fsig->params [1]->type == MONO_TYPE_R4);
 
 		gboolean indirect = FALSE;
+		int elems = 4, index = args [1]->inst_c0;
 		int dreg = load_simd_vreg (cfg, cmethod, args [0], &indirect);
 
-		MONO_EMIT_NEW_BIALU_IMM (cfg, OP_COMPARE_IMM, -1, args [1]->dreg, len);
-		MONO_EMIT_NEW_COND_EXC (cfg, GE_UN, "ArgumentOutOfRangeException");
-		ins = emit_simd_ins (cfg, klass, OP_XINSERT_R4, dreg, args [2]->dreg);
-		ins->sreg3 = args [1]->dreg;
+		// Bounds check only if the index is out of range
+		if (index < 0 || index >= elems) {
+			MONO_EMIT_NEW_BIALU_IMM (cfg, OP_COMPARE_IMM, -1, args [1]->dreg, len);
+			MONO_EMIT_NEW_COND_EXC (cfg, GE_UN, "ArgumentOutOfRangeException");
+		}
+
+		if (COMPILE_LLVM (cfg)) {
+			ins = emit_simd_ins (cfg, klass, OP_XINSERT_R4, dreg, args [2]->dreg);
+			ins->sreg3 = args [1]->dreg;
+		} else {
+			ins = emit_simd_ins (cfg, klass, OP_INSERT_R4, dreg, args [2]->dreg);
+			ins->inst_c0 = index;
```

Comment:
`set_Item` can be called with nonconstant `index`, in which case I think the mini path would not work. You may want to look at https://github.com/dotnet/runtime/blob/a0a8ce681afda6b662e450ac9532b0d4c125520c/src/mono/mono/mini/simd-intrinsics.c#L2183C11-L2183C11, where the constant case is detected (an optimized if possible). If `index` is not constant, you'll need to use `OP_XINSERT_R4`

Reply:
Agree with you. 
Updated the PR.

Reply:
This look OK now.